Sumo-Logic-Tools/Event_Based_S3_Automation/script.py

```diff
@@ -0,0 +1,265 @@
+	mapOfBucketNameToEndpoint = defaultdict(list)
+
+	def __init__(self, base_url, access_id, access_key):
+		self.base_url = base_url
+		self.access_id = access_id
+		self.access_key = access_key
+
+	def calculateMappingOfBucketNameToEndpoint(self):
+		listOfCollectors = self.getListOfCollectors()
+
+		listOfSourcesLink = self.getSourcesLinks(listOfCollectors)
+
+		listOfPollingSources = self.getPollingSources(listOfSourcesLink)
+
+		self.populateMapOfBucketNameToEndpoint(listOfPollingSources)
+
+		return self.mapOfBucketNameToEndpoint
+
+	def getListOfCollectors(self):
+		return self.makeGetRequest(self.collector_url).json()
+
+	def makeGetRequest(self, url):
+		return requests.get(url = self.base_url + str(url), auth = HTTPBasicAuth(self.access_id, self.access_key))
+
+	def getSourcesLinks(self, listOfCollectors):
+		listOfSourcesLink = list()
+		try:
+			for collector in listOfCollectors['collectors']:
+				links = collector['links']
+				listOfSourcesLink.extend(self.extractSourcesLink(collector['links']))
+		except KeyError as e:
+			try:
+				collector = listOfCollectors['collector']
+				links = collector['links']
+				listOfSourcesLink.extend(self.extractSourcesLink(collector['links']))
+			except KeyError as e:
+				print("The API did not return any collectors!")
+		return listOfSourcesLink
+
+	def extractSourcesLink(self, links):
+		listOfSourcesLink = list()
+		for link in links:
+			if (link['rel'] == 'sources'):
+				listOfSourcesLink.append(link['href'])
+
+		return listOfSourcesLink
+
+	def getPollingSources(self, listOfSourcesLink):
+		listOfPollingSources = list()
+		for sourceLink in listOfSourcesLink:
+			allSources = self.makeGetRequest(sourceLink).json()
+			for source in allSources['sources']:
+				if (self.isPollingSource(source) == True):
+					listOfPollingSources.append(source)
+
+		return listOfPollingSources
+
+	def isPollingSource(self, source):
+		# get returns None if contentType is not available. The source cannot be polling if it does not have a content type
+		contentType = source.get('contentType')
+		return (contentType in self.pollingTypeSources)
+
+	def populateMapOfBucketNameToEndpoint(self, listOfPollingSources):
+		for source in listOfPollingSources:
+			bucketName = self.getBucketName(source)
+			endpoint = self.getEndpoint(source)
+			if (bucketName is not None and endpoint is not None):
+				self.mapOfBucketNameToEndpoint[bucketName].append(endpoint)
+
+	def getBucketName(self, source):
+		try:
+			thirdPartyRef = source['thirdPartyRef']
+			resources = thirdPartyRef['resources']
+			path = resources[0]['path']
+			return path['bucketName']
+		except:
+			return None
+
+	def getEndpoint(self, source):
+		return source.get('url')
+
+
+'''
+Helps in setting up event based S3 using the appropriate AWS APIs.
+It does the following tasks:
+1. Using the map of bucket name to endpoint generated above, it creates a topic per bucket. (So a topic for every key in that map)
```

Comment:
Can/should we optimise this based on path expressions so that we don't get notifications for an object multiple times? Basing this on the mail thread `Re: FW: Sumo Surgery - S3 event-based notification beta` (I would have linked to it, but I don't think there's a way to link to email threads as of now 🤔)

Reply:
@jksingh has recommended that if we want to implement this, it should be in v1 since it's unlikely that people will prune their configurations later

Reply:
Yes. This would require some time and thought. I was thinking of creating a separate PR for it so that it is easier to discuss and review. I agree this capability should be added before we encourage customers to start using this script. 
The initial purpose of this script was to provide a quick solution for Braintree only. Hence we had decided to leave out the optimisation.

Reply:
Let's look at this aspect in a separate PR. The logic change will be clearer and simpler to review in isolation.